llvm/lib/Support/KnownBits.cpp

```diff
@@ -403,8 +391,7 @@ KnownBits KnownBits::lshr(const KnownBits &LHS, const KnownBits &RHS,
   if (Exact) {
     unsigned FirstOne = LHS.countMaxTrailingZeros();
     if (FirstOne < MinShiftAmount) {
-      // Always poison. Return zero because we don't like returning conflict.
-      Known.setAllZero();
+      // Always poison.
```

Comment:
Similar here and in the ashr implementation. We go from returning zero to returning unknown, which is responsible for the regressions in SCEV tests where we go from range `[0, 1)` to `full-set`.

The correct replacement would be something like setConflict(), but given that we don't have that yet and wouldn't be able to make use of it for folding I think it's best to leave these cases alone.

Reply:
Hm, I see. In that case I think it would be best to just revert *all* the setAllZero() related changes. We will have to update consuming code first so it can do something useful with conflict values before we start generating them.